File: Commons/Math/src/main/java/com/greentree/common/math/vector/ActionVector2f.java
package com.greentree.common.math.vector;

import java.util.function.Consumer;

import com.greentree.action.observer.object.EventAction;

public class ActionVector2f extends AbstractVector2f implements VectorAction {

	private boolean actionX, actionY;

	private final EventAction<ActionVector2f> action = new EventAction<>();
	private final ActionFloat x, y;

	private static final long serialVersionUID = 1L;

	public ActionVector2f(float x, float y) {
		this.x = new ActionFloat();
		this.x.set(x);
		this.x.addListener(l -> actionX = true);
		
		this.y = new ActionFloat();
		this.y.set(y);
		this.y.addListener(l -> actionY = true);
	}
	public ActionVector2f(float f) {
		this(f, f);
	}
	public ActionVector2f() {
		this(0);
	}
	
	public void addListener(Consumer<ActionVector2f> l){
		action.addListener(l);
	}

	@Override
	public void x(float x) {
		this.x.set(x);
	}

	@Override
	public void y(float y) {
		this.y.set(y);
	}

	public final void tryAction(){
		if(actionX || actionY) {
			actionX = false;
			actionY = false;
			action.event(this);
		}
	}

	@Override
	public float x() {
		tryAction();
		return x.get();
	}

	@Override
	public float y() {
		tryAction();
		return y.get();
	}

}

File: Commons/Math/src/main/kotlin/com/greentree/commons/math/vector/ActionVector2f.kt
package com.greentree.commons.math.vector

import com.greentree.commons.action.observer.`object`.EventAction
import java.util.function.Consumer

class ActionVector2f : AbstractMutableVector2f {

    override var x: Float
        get(){
            tryAction()
            return X.get()
        }
        set(value) = X.set(value)
    override var y: Float
        get() {
            tryAction()
            return X.get()
        }
        set(value) = Y.set(value)

    private var actionX = false
    private var actionY = false
    private val action: EventAction<ActionVector2f> = EventAction<ActionVector2f>()
    private val X: ActionFloat
    private val Y: ActionFloat

    @JvmOverloads
    constructor(value: Float = 0f) : this(value, value)

    constructor(x: Float, y: Float) {
        X = ActionFloat(x)
        Y = ActionFloat(y)
        this.X.addListener(Consumer { _ -> actionX = true })
        this.Y.addListener(Consumer { _ -> actionY = true })
    }

    fun addListener(l: Consumer<ActionVector2f?>?) {
        action.addListener(l)
    }

    fun tryAction() {
        if (actionX || actionY) {
            actionX = false
            actionY = false
            action.event(this)
        }
    }

}